[WIP] Docs: Add initial draft of Strands documentation (feedback wanted)

Abhayaj247 opened the pull request:
**Note:** This PR replaces #7920 to resolve merge conflicts that occurred while rebasing onto the `dev-2.0` branch.
***

This PR introduces new JSDoc reference entries for several `p5.strands` hooks, as discussed in #7919 and based on recent feedback.

**Key updates:**
- Added detailed documentation and JavaScript-only usage examples for the following hooks:
  - `getWorldInputs`
  - `combineColors`
  - `getPointSize`
- Described callback signatures for each hook.
- Placed the new doc comments in `src/webgl/p5.strands.js` for clarity and modularity.

**Request for feedback:**
- Please review the documentation style, level of detail, and the chosen file location.
- If another location is preferred for these comments, I’m happy to move them.

Once I receive feedback and confirmation on the format and placement, I’ll proceed to document the remaining hooks in the same style.

Thank you for your time and guidance!

Abhayaj247 commented:
**Hi @perminder-17  Could you please review this new PR? It replaces the previous one #7920 due to merge conflicts and now targets `dev-2.0`.** 

Thank you!

davepagurek commented on src/webgl/ShaderGenerator.js:
```diff
@@ -1638,3 +1638,159 @@ export default shadergenerator;
+ * Registers a callback to modify the world-space properties of each vertex in a shader. This hook can be used inside {@link p5.baseMaterialShader}.modify() and similar shader modify calls to customize vertex positions, normals, texture coordinates, and colors before rendering. "World space" refers to the coordinate system of the 3D scene, before any camera or projection transformations are appli ...
+ *
+ * This hook is available in:
+ * - {@link p5.baseMaterialShader}
```
I think although this is valid JSDoc, we don't yet support it in p5's jsdoc-to-reference pipeline. Currently we use `<a href="...">`, e.g.: https://github.com/processing/p5.js/blob/0da04307caf28306f55215c05991d35522535968/src/type/p5.Font.js#L1207

Abhayaj247 replied:
Hi @davepagurek ,
I've updated all references to use the <a href="..."> syntax that p5.js supports, following the pattern from p5.Font.js. All links now use the format <a href="#/p5/baseMaterialShader">baseMaterialShader()</a>.

davepagurek commented on src/webgl/ShaderGenerator.js:
```diff
@@ -1638,3 +1638,159 @@ export default shadergenerator;
+
+/* ------------------------------------------------------------- */
+/**
+ * @typedef {Object} Vertex
```
I'm also not sure that typedef docs will show up in the reference, even though this does simplify the writing of the function parameter. If you [test your branch on the p5.js-website repo](https://github.com/processing/p5.js-website/blob/main/docs/scripts.md#testing-the-docs-of-your-fork), what does this look like? If it doesn't render, we may need to either document this object inline, or update the website to be able to support this syntax.

Abhayaj247 replied:
@davepagurek Thank you for the guidance on the @typedef issue. I've removed it and documented the vertex object properties inline in the parameter description to ensure clarity and compatibility with the documentation pipeline.

davepagurek commented on src/webgl/ShaderGenerator.js:
```diff
@@ -1638,3 +1638,159 @@ export default shadergenerator;
+ * @param {function(ColorComponents): { color: {r: number, g: number, b: number}, opacity: number }} callback
+ *        A callback function which receives a ColorComponents struct and returns the final color and opacity.
+ *
+ * @see {@link p5.baseMaterialShader}
```
Similar to above, we probably don't support `@see` yet, but feel free to use regular links in a paragraph explaining why you might want to check out these related references!

Abhayaj247 replied:
@davepagurek , I've removed all @see tags since they're not supported in the current documentation pipeline. The documentation now relies on clear descriptions and contextual examples to guide users to related references.

davepagurek commented on src/webgl/ShaderGenerator.js:
```diff
@@ -1638,3 +1638,159 @@ export default shadergenerator;
+ *   myShader = baseMaterialShader().modify(() => {
+ *     combineColors(components => {
+ *       // Custom color combination: add a red tint
+ *       let color = {
```
Although you can *access* `.r`, `.g`, `.b`, and `.a` properties, the return value would need to be either `[r, g, b, a]` as an array, or more explicitly, `vec4(r, g, b, a)` -- we don't yet support an object notation for this. (Also it will need to be just the one object, and not an object with separate color and opacity properties, for it to compile correctly I think.)

davepagurek replied:
This also makes me wonder what the best way to document the `inputs` object is, since it's not exactly any normal type. @lukeplowden do you have any thoughts on this? We could call it a `vec4` and then explain in the description of the function what this means, or even not give it a type at all, and just have an explanation in the description.

davepagurek commented on src/webgl/ShaderGenerator.js:
```diff
@@ -1700,19 +1689,16 @@ if (typeof p5 !== 'undefined') {
  * @experimental
  * @description
- * Registers a callback to customize how color components are combined in the fragment shader. This hook can be used inside {@link p5.baseMaterialShader}.modify() and similar shader modify calls to control the final color output of a material. The callback receives a ColorComponents struct and should return an object with a `color` property ({ r, g, b }) and an `opacity` property (number).
+ * Registers a callback to customize how color components are combined in the fragment shader. This hook can be used inside <a href="#/p5/baseMaterialShader">baseMaterialShader()</a>.modify() and similar shader modify calls to control the final color output of a material. The callback receives color components (baseColor, diffuse, ambientColor, ambient, specularColor, specular, emissive, opacity) ...
```
Can we describe the input as being an object with the properties `baseColor`, `diffuse`, `ambientColor`, etc? I think it would be important to describe the size of each item (e.g. position/normal are vec3s, texture coordinates are vec2s, color is a vec4) -- although maybe rather than using terms like `vec3` we could describe it as "a vector with three components" to avoid referencing the `vec3` type, since there aren't user-facing docs for that yet. A bullet list might work if that's too much info to put in one sentence.

One other note is that when we refer to a name of a property/variable/etc in code, such as when listing properties here, we should put the names in backticks so they render as code (`baseColor` instead of baseColor.)

Abhayaj247 replied:
```
/**
 * @function combineColors
 * @experimental
 * @description
 * Registers a callback to customize how color components are combined in the fragment shader. This hook can be used inside <a href="#/p5/baseMaterialShader">baseMaterialShader()</a>.modify() and similar shader modify calls to control the final color output of a material. The callback receives an object with the following properties:
 *
 * - `baseColor`: a vector with three components representing the base color (red, green, blue)
 * - `diffuse`: a single number representing the diffuse reflection
 * - `ambientColor`: a vector with three components representing the ambient color
 * - `ambient`: a single number representing the ambient reflection
 * - `specularColor`: a vector with three components representing the specular color
 * - `specular`: a single number representing the specular reflection
 * - `emissive`: a vector with three components representing the emissive color
 * - `opacity`: a single number representing the opacity
 *
 * The callback should return a vector with four components (red, green, blue, alpha) for the final color.
 *
 * This hook is available in:
 * - <a href="#/p5/baseMaterialShader">baseMaterialShader()</a>
 * - <a href="#/p5/baseNormalShader">baseNormalShader()</a>
 * - <a href="#/p5/baseColorShader">baseColorShader()</a>
 * - <a href="#/p5/baseStrokeShader">baseStrokeShader()</a>
 * - <a href="#/p5/baseFilterShader">baseFilterShader()</a>
 *
 * @param {function} callback
 *        A callback function which receives the object described above and returns a vector with four components for the final color.
 *
 * @example
 * <div modernizr='webgl'>
 * <code>
 * let myShader;
 * function setup() {
 *   createCanvas(200, 200, WEBGL);
 *   myShader = baseMaterialShader().modify(() => {
 *     combineColors(components => {
 *       // Custom color combination: add a red tint
 *       let r = components.baseColor[0] * components.diffuse +
 *               components.ambientColor[0] * components.ambient +
 *               components.specularColor[0] * components.specular +
 *               components.emissive[0] + 0.2;
 *       let g = components.baseColor[1] * components.diffuse +
 *               components.ambientColor[1] * components.ambient +
 *               components.specularColor[1] * components.specular +
 *               components.emissive[1];
 *       let b = components.baseColor[2] * components.diffuse +
 *               components.ambientColor[2] * components.ambient +
 *               components.specularColor[2] * components.specular +
 *               components.emissive[2];
 *       return [r, g, b, components.opacity];
 *     });
 *   });
 * }
 * function draw() {
 *   background(255);
 *   shader(myShader);
 *   lights();
 *   noStroke();
 *   fill('red');
 *   sphere(50);
 * }
 * </code>
 * </div>
 */
 
 ```
 
Thank you for your feedback @davepagurek ! I’ve updated the documentation for ```combineColors``` to describe the input as an object with each property listed in a bullet list, using backticks for property names and describing the size of each in plain language. The example now uses array notation for vector properties. Please let me know if you’d like any further adjustments.

Abhayaj247 replied:
```
/**
 * @function combineColors
 * @experimental
 * @description
 * Registers a callback to customize how color components are combined in the fragment shader. This hook can be used inside <a href="#/p5/baseMaterialShader">baseMaterialShader()</a>.modify() and similar shader modify calls to control the final color output of a material. The callback receives an object with the following properties:
 *
 * - `baseColor`: a vector with three components representing the base color (red, green, blue)
 * - `diffuse`: a single number representing the diffuse reflection
 * - `ambientColor`: a vector with three components representing the ambient color
 * - `ambient`: a single number representing the ambient reflection
 * - `specularColor`: a vector with three components representing the specular color
 * - `specular`: a single number representing the specular reflection
 * - `emissive`: a vector with three components representing the emissive color
 * - `opacity`: a single number representing the opacity
 *
 * The callback should return a vector with four components (red, green, blue, alpha) for the final color.
 *
 * This hook is available in:
 * - <a href="#/p5/baseMaterialShader">baseMaterialShader()</a>
 * - <a href="#/p5/baseNormalShader">baseNormalShader()</a>
 * - <a href="#/p5/baseColorShader">baseColorShader()</a>
 * - <a href="#/p5/baseStrokeShader">baseStrokeShader()</a>
 * - <a href="#/p5/baseFilterShader">baseFilterShader()</a>
 *
 * @param {function} callback
 *        A callback function which receives the object described above and returns a vector with four components for the final color.
 *
 * @example
 * <div modernizr='webgl'>
 * <code>
 * let myShader;
 * function setup() {
 *   createCanvas(200, 200, WEBGL);
 *   myShader = baseMaterialShader().modify(() => {
 *     combineColors(components => {
 *       // Custom color combination: add a red tint
 *       let r = components.baseColor[0] * components.diffuse +
 *               components.ambientColor[0] * components.ambient +
 *               components.specularColor[0] * components.specular +
 *               components.emissive[0] + 0.2;
 *       let g = components.baseColor[1] * components.diffuse +
 *               components.ambientColor[1] * components.ambient +
 *               components.specularColor[1] * components.specular +
 *               components.emissive[1];
 *       let b = components.baseColor[2] * components.diffuse +
 *               components.ambientColor[2] * components.ambient +
 *               components.specularColor[2] * components.specular +
 *               components.emissive[2];
 *       return [r, g, b, components.opacity];
 *     });
 *   });
 * }
 * function draw() {
 *   background(255);
 *   shader(myShader);
 *   lights();
 *   noStroke();
 *   fill('red');
 *   sphere(50);
 * }
 * </code>
 * </div>
 */
 
 ```
 
> @davepagurek Thank you for your feedback! I’ve updated the documentation for ```combineColors``` to describe the input as an object with each property listed in a bullet list, using backticks for property names and describing the size of each in plain language. The example now uses array notation for vector properties. Please let me know if you’d like any further adjustments.

davepagurek commented on src/webgl/ShaderGenerator.js:
```diff
@@ -1638,3 +1638,141 @@ export default shadergenerator;
+ * Registers a callback to customize how color components are combined in the fragment shader. This hook can be used inside <a href="#/p5/baseMaterialShader">baseMaterialShader()</a>.modify() and similar shader modify calls to control the final color output of a material. The callback receives color components (baseColor, diffuse, ambientColor, ambient, specularColor, specular, emissive, opacity) ...
+ *
+ * This hook is available in:
+ * - <a href="#/p5/baseMaterialShader">baseMaterialShader()</a>
```
Looks great!

Abhayaj247 replied:
@davepagurek Thank you! I appreciate your guidance throughout this process. If there’s anything else you’d like me to adjust or clarify, please let me know.

davepagurek commented on src/webgl/ShaderGenerator.js:
```diff
@@ -1638,3 +1638,141 @@ export default shadergenerator;
+ */
+
+/**
+ * @function getPointSize
```
Just a heads up -- currently the only public base shaders are:

- [baseColorShader()](https://beta.p5js.org/reference/p5/baseColorShader)
- [baseFilterShader()](https://beta.p5js.org/reference/p5/baseFilterShader)
- [baseMaterialShader()](https://beta.p5js.org/reference/p5/baseMaterialShader)
- [baseNormalShader()](https://beta.p5js.org/reference/p5/baseNormalShader)
- [baseStrokeShader()](https://beta.p5js.org/reference/p5/baseStrokeShader)

The point shader, while it has some hooks, is currently still private, as we may end up using the stroke shader to render points in the future.

davepagurek replied:
I think `getPointSize` itself is only present in the base point shader, so I think we don't need this one.

Abhayaj247 commented:
@davepagurek Thank you for your thoughtful feedback and guidance throughout this process. I have addressed all the points you raised:

- The documentation for combineColors now describes the input as an object with each property listed in a bullet list, using backticks for property names and describing the size of each in plain language. The example uses array notation for vector properties.
- The documentation for getPointSize has been updated to reference only the public base shaders, with a clear and readable format.
- All other formatting and clarity suggestions have been incorporated as well.

Please see the updated code blocks above for reference.

If you have any further suggestions or would like additional adjustments, I’m happy to make them!

Abhayaj247 commented:
> Thanks @Abhayaj247, the docs are looking good! I think maybe we're ready to give the other hooks the same treatment now.

@davepagurek Thank you so much for your guidance and support throughout this process! I’m glad the documentation updates meet your expectations. I’ll proceed to document the remaining hooks in the same style. 
Please let me know if you have any specific preferences or additional feedback as I work on those.

Abhayaj247 commented:
**Hi @davepagurek ,**

**As I continue updating the shader documentation, I’ve reviewed the codebase and found these hooks as the remaining ones that appear to be public or user-facing:**

- **`beforeVertex`**
- **`afterVertex`**
- **`beforeFragment`**
- **`afterFragment`**
- **`getFinalColor`**
- **`shouldDiscard`**
- **`getLocalNormal`**
- **`getWorldNormal`**
- **`getUV`**
- **`getVertexColor`**

**Additionally, I noticed hooks like `getLinePosition`, `getLineCenter`, and `getStrokeWeight` in the codebase, but they aren’t referenced in the main documentation. My understanding is that these are likely internal, but please let me know if you’d like them included as well.**

**Could you please confirm that this list covers all the hooks intended for user documentation? Once confirmed, I’ll proceed to document them in the new style.**

**Thank you for your guidance!**

davepagurek commented:
Hi @Abhayaj247! I think some of those, e.g. `getUV`, are from 1.x using an earlier beta of the hooks API, and are no longer present in the 2.0 branch. The ones to document are the ones in the tables on these pages on the beta.p5js.org 2.0 reference:

- [baseColorShader()](https://beta.p5js.org/reference/p5/baseColorShader)
- [baseFilterShader()](https://beta.p5js.org/reference/p5/baseFilterShader)
- [baseMaterialShader()](https://beta.p5js.org/reference/p5/baseMaterialShader)
- [baseNormalShader()](https://beta.p5js.org/reference/p5/baseNormalShader)
- [baseStrokeShader()](https://beta.p5js.org/reference/p5/baseStrokeShader)

Abhayaj247 commented:
**Hi @davepagurek** 

**I’ve now added documentation for the remaining hooks listed in the tables for the following shaders in the 2.0 reference :**

- **[`baseColorShader`](https://beta.p5js.org/reference/p5/basecolorshader/)**
- **[`baseFilterShader`](https://beta.p5js.org/reference/p5/basefiltershader/)**
- **[`baseMaterialShader`](https://beta.p5js.org/reference/p5/basematerialshader/)**
- **[`baseNormalShader`](https://beta.p5js.org/reference/p5/basenormalshader/)**
- **[`baseStrokeShader`](https://beta.p5js.org/reference/p5/basestrokeshader/)**

**The newly documented hooks are :**

- **`beforeVertex`**
- **`getObjectInputs`**
- **`getCameraInputs`**
- **`afterVertex`**
- **`beforeFragment`**
- **`getPixelInputs`**
- **`shouldDiscard`**
- **`getFinalColor`**
- **`afterFragment`**
- **`getColor`**

**The hooks `getWorldInputs` and `combineColors` were already documented in previous updates.**

**Please let me know if there are any further adjustments or if you’d like me to include any additional hooks.** 

**Thank you again for your guidance!**

Abhayaj247 commented on src/webgl/ShaderGenerator.js:
```diff
@@ -1638,3 +1638,439 @@ export default shadergenerator;
+ *   myShader = baseColorShader().modify(() => {
+ *     getObjectInputs(inputs => {
+ *       // Raise all vertices by 10 units
+ *       inputs.position[1] += 10;
```
> Similar to before, we'd need to use `.y` here. It might also be useful to change the y value based on the x value, e.g. to make a sine wave along an object, rather than just a constant offset, as it might not be obvious visually what the change has been for that.

```
getObjectInputs(inputs => {
  // Create a sine wave along the x axis in object space
  inputs.position.y += 20 * sin(inputs.position.x * 0.1 + millis() * 0.002);
  return inputs;
});
```
@davepagurek , Changed all vector accesses to use `.x`, `.y`, `.z` etc., and updated the example to use a sine wave for a visually clear effect.

Abhayaj247 commented on src/webgl/ShaderGenerator.js:
```diff
@@ -1638,3 +1638,439 @@ export default shadergenerator;
+ *   myShader = baseColorShader().modify(() => {
+ *     beforeVertex(() => {
+ *       // Set up a variable for later use
+ *       let wave = sin(millis() * 0.001);
```
> I think to have a variable for later use within the same shader, we need support for shader-level globals. We currently have varyings only, which are for passing data between the vertex and fragment shader, where the before/after hooks aren't super useful. Maybe we don't need examples for before/after until we have real support for those? @lukeplowden do you know if we have an issue for that yet?

```
/**
 * @function beforeVertex
 * ...
 * Note: This hook is currently limited to per-vertex operations; storing variables for later use is not supported.
 */
 ```
 @davepagurek ,Removed the `@example` blocks for `beforeVertex` and `afterVertex` hooks, and left a note about current limitations.

Abhayaj247 commented on src/webgl/ShaderGenerator.js:
```diff
@@ -1638,3 +1638,439 @@ export default shadergenerator;
+ * @function getObjectInputs
+ * @description
+ * Registers a callback to modify the properties of each vertex before any transformations are applied in the vertex shader. This hook can be used inside <a href="#/p5/baseColorShader">baseColorShader()</a>.modify() and similar shader modify calls to move, color, or otherwise modify the raw model data. The callback receives an object with the following properties:
+ * - `position`: `[x, y, z]` — the original position of the vertex
```
> Could we continue using the same "a vector with three components" terminology from above everywhere?

```
 * - `position`: a vector with three components representing the position after camera transformation
 * - `normal`: a vector with three components representing the normal after camera transformation
 * - `texCoord`: a vector with two components representing the texture coordinates
 * - `color`: a vector with four components representing the color of the vertex
 ```
@davepagurek ,Updated all doc blocks to consistently use the phrase “a vector with three components” (or “four components”, etc.) for all relevant properties, as requested.

Abhayaj247 commented on src/webgl/ShaderGenerator.js:
```diff
@@ -1638,3 +1638,439 @@ export default shadergenerator;
+ *   myShader = baseFilterShader().modify(() => {
+ *     getColor((inputs, canvasContent) => {
+ *       // Return the original color
+ *       return getTexture(canvasContent, inputs.texCoord);
```
> It might be nice to do some kind of change here, e.g. warping the coordinate or changing the color

```
getColor((inputs, canvasContent) => {
  // Warp the texture coordinates for a wavy effect
  let warped = [inputs.texCoord.x, inputs.texCoord.y + 0.1 * sin(inputs.texCoord.x * 10.0)];
  return getTexture(canvasContent, warped);
});
```
@davepagurek ,Updated the `getColor` example to warp the coordinates for a visible effect.

Abhayaj247 commented:
> **Thank you for the detailed feedback @davepagurek ! I have addressed all the points as follows :**
> 
> **1. Visual Examples :**
> 
> * All code examples now include a `draw()` function that draws a shape and applies the shader, so the effect of each hook is visually demonstrated.
> 
> **2. Vector Property Access :**
> 
> * All vector and color accesses now use property notation (e.g., `.y`, `.r`, `.g`, `.b`) instead of array indices, as requested.
> 
> **3. Sine Wave Example :**
> 
> * The `getObjectInputs` example now uses a sine wave along the `x` axis for a visually clear effect, rather than a constant offset.
> 
> **4. before/afterVertex Examples :**
> 
> * The `beforeVertex` and `afterVertex` hooks no longer have examples, per your suggestion, and include a note about current limitations.
> 
> **5. Consistent Vector Terminology :**
> 
> * All doc blocks use the consistent phrase “a vector with three components” (or “four components”, etc.) for all relevant properties, as requested.
> 
> **6. Bullet Lists :**
> 
> * All callback object properties are now listed as bullet points in the documentation for clarity.
> 
> **7. getColor Example :**
> 
> * The `getColor` example now warps the coordinates for a visible effect, rather than just returning the original color.
> 
> **8. Position Examples in Input Hooks :**
> 
> * The `getObjectInputs`, `getWorldInputs`, and `getCameraInputs` examples all include a position modification, with comments explaining the coordinate space difference.
> 
> **If you have any further suggestions, please let me know!**
> 
> **Thank you again for your guidance.**


**Hi @davepagurek ,**

**Just a gentle reminder regarding my previous comment about addressing all the feedback points for the shader hooks documentation.**

**I've made all the requested changes including visual examples, vector property access updates, sine wave examples, and consistent terminology throughout the documentation.**

**If you have a moment, could you please review and let me know if any further adjustments are needed?**

**Thank you for your time!**

lukeplowden commented:
Hey @Abhayaj247, we are aware of the issue, there's no need to keep bumping it! You can tag me instead of Dave, if you must. I will give it a read over later today.

davepagurek commented on src/webgl/ShaderGenerator.js:
```diff
@@ -1638,3 +1638,480 @@ export default shadergenerator;
+ *     });
+ *     getFinalColor(color => {
+ *       // Use the value set in beforeFragment to tint the color
+ *       color[0] *= this.brightness; // Tint red channel
```
Looks like we still have some uses of array indexing. This should be `color.r *= ...`

davepagurek replied:
(Although maybe don't worry about this particular instance as I mentioned earlier that we probably want to `@private` this particular method.)

davepagurek commented on src/webgl/ShaderGenerator.js:
```diff
@@ -1638,3 +1638,480 @@ export default shadergenerator;
+ *   myShader = baseMaterialShader().modify(() => {
+ *     getPixelInputs(inputs => {
+ *       // Animate opacity based on x position
+ *       inputs.opacity = 0.5 + 0.5 * sin(inputs.texCoord.x * 10.0 + millis() * 0.002);
```
In order to use `millis`, we currently have to pass it in via a uniform:
```js
let t = uniformFloat(() => millis());
getPixelInputs(inputs => {
  inputs.opacity = 0.5 + 0.5 * sin(inputs.texCoord.x * 10 + t * 0.002);
  return inputs;
})
```

davepagurek commented on src/webgl/ShaderGenerator.js:
```diff
@@ -1638,3 +1638,480 @@ export default shadergenerator;
+ *   myShader = baseStrokeShader().modify(() => {
+ *     shouldDiscard(willDiscard => {
+ *       // Discard fragments outside a circular region
+ *       return willDiscard || (inputs.position.x * inputs.position.x + inputs.position.y * inputs.position.y > 2500.0);
```
`inputs` isn't defined here. Since we'd have to pass values between `getPixelInputs` and this, maybe this method is also one that we should `@private` until we have support for that.

davepagurek commented on src/webgl/ShaderGenerator.js:
```diff
@@ -1638,3 +1638,480 @@ export default shadergenerator;
+ *   shader(myShader);
+ *   noStroke();
+ *   fill('orange');
+ *   box(100);
```
Since a box doesn't have that many vertices, only the corners will rotate, so you won't see the sine wave as well. it might be more obvious to use a `sphere` here

Abhayaj247 replied:
I’ve updated the example to use `sphere` instead of `box` to better demonstrate the effect.

davepagurek replied:
Just took a look again, seems like it's still a box?

Abhayaj247 replied:
Thanks for catching that - looks like I missed committing the change in the last update. I’ve now replaced box(100) with sphere(100) in the example so the sine wave effect is more visible.

lukeplowden commented:
> * Currently the ShaderGenerator file has `@submodule ShaderGenerator` at the top. The submodule corresponds to the subheading in the reference index. I think it might make sense to do `@submodule Material` to put these in the same Material subheading as other p5 reference items. (Thoughts, @lukeplowden?)

That sounds good to me, as ShaderGenerator is a temporary name. 

Let me know if you have any issues building the website repo @Abhayaj247

Abhayaj247 commented:
> > * Currently the ShaderGenerator file has `@submodule ShaderGenerator` at the top. The submodule corresponds to the subheading in the reference index. I think it might make sense to do `@submodule Material` to put these in the same Material subheading as other p5 reference items. (Thoughts, @lukeplowden?)
> 
> That sounds good to me, as ShaderGenerator is a temporary name.
> 
> Let me know if you have any issues building the website repo @Abhayaj247

Summary of Build Process and Changes in p5.js-website:

Hi @lukeplowden , here’s an overview of my build and testing process:

1. Environment Variable Management 

- Initially, I encountered errors with Unix-style environment variable syntax `(PUBLIC_P5_LIBRARY_PATH=...)` not working on Windows.
- To resolve this, I switched to using a `.env` fileand the `dotenv` package for environment variables, ensuring cross-platform compatibility.

2. Frontmatter Reference Errors

- During the build, I encountered `InvalidContentEntryFrontmatterError` for several `.mdx` tutorial files due to invalid references (e.g., `en/p5/preload`, `en/p5sound/p5oscillator`).
- I removed all invalid references from the frontmatter in the affected files, resolving the build errors.

3. Testing and Verification

- After committing and pushing my changes to the `update-strands-docs` branch, I ran:

`npm run custom:dev https://github.com/Abhayaj247/p5.js.git#update-strands-docs`

in the `p5.js-website` repo.

4. Previous Errors Encountered

- Environment variable not recognized on Windows.
- Missing `dotenv` package.
- Invalid frontmatter references causing build failures.


**However, I’ve committed and pushed my changes, and rebuilt the website using the recommended workflow. However, I’m having trouble locating my updates and example sketches in the website documentation. Could you please advise which section, page, or search terms I should use to view my changes? Thank you!"**

Abhayaj247 commented:
**Hi everyone, thank you all for the helpful feedback!**

**I’ve implemented all the suggested changes so far:**

**- Replaced array indexing with property access like `color.r`, `color.g`, `color.a` where applicable.**
**- Updated `millis()` usage to be passed via a uniform, following the recommended pattern.**
**- Marked methods such as `beforeVertex`, `afterVertex`, `beforeFragment`, and `afterFragment` as `@private` since variable passing between shader sections isn’t supported yet.**
**- Simplified code examples (e.g., switched from `ellipse()` to `circle()`, and from `box()` to `sphere()` for better visual clarity).**
**- Updated `@submodule` to `Material` instead of `ShaderGenerator` to align with the rest of the reference structure.**

**I've committed and pushed these changes, and rebuilt the site using `npm run custom:dev` in the `p5.js-website` repo. The build succeeds without errors, but I’m still having trouble locating the new entries and example sketches in the Reference section.**

**Could you please confirm which page, section, or search terms I should use to verify whether everything is showing up as expected? Thanks again for your guidance!**

Abhayaj247 commented:
Thanks @davepagurek ! That’s really helpful. I hadn’t switched to the 2.0 branch of the p5.js-website repo before running the build, so I think that’s why I didn’t see it. I’ll follow the same steps and verify that all the examples are rendering correctly and clearly demonstrate the intended behaviors.

Abhayaj247 commented:
Hi @davepagurek , Just a quick update on the status of the examples after rebuilding the site using the recommended steps:

| Hook | Example Status | Notes |
|------|----------------|-------|
| `getWorldInputs` | ✅ Appearing | Under **3D > Material** |
| `combineColors` | ❌ Not appearing | |
| `getPixelInputs` | ❌ Not appearing | |
| `getFinalColor` | ✅ Appearing | But **not under 3D** section |
| `getColor` | ✅ Appearing | Also **not under 3D** |
| `shouldDiscard` | ❌ Not appearing | |
| `getObjectInputs` | ⚠️ Appearing incorrectly | Doesn’t display as expected |
| `getCameraInputs` | ✅ Appearing | Under **3D** section |
| `beforeVertex`, `afterVertex`, `beforeFragment`, `afterFragment` | 🔒 Marked as `@private` | As recommended earlier |

Abhayaj247 commented:
Hi @davepagurek , Just a quick update on the status of the examples after rebuilding the site using the recommended steps:

| Hook | Example Status | Notes |
|------|----------------|-------|
| `getWorldInputs` | ✅ Appearing | Under **3D > Material** |
| `combineColors` | ❌ Not appearing | |
| `getPixelInputs` | ❌ Not appearing | |
| `getFinalColor` | ✅ Appearing | But **not as a 3D**  |
| `getColor` | ✅ Appearing | But **not as a 3D** |
| `shouldDiscard` | ❌ Not appearing | |
| `getObjectInputs` | ⚠️ Appearing incorrectly | Doesn’t display as expected |
| `getCameraInputs` | ✅ Appearing | Under **3D > Material** |
| `beforeVertex`, `afterVertex`, `beforeFragment`, `afterFragment` | 🔒 Marked as `@private` | As recommended earlier |


I'm attaching screenshots and short video clips below to show how each hook appears (or doesn't) in the local build:

---

#### 📸 `getWorldInputs`
_Attached Screenshot/Video below:_
![getWorldInputs](http://localhost:4321/reference/p5/getWorldInputs/)

---

davepagurek commented on src/webgl/ShaderGenerator.js:
```diff
@@ -1638,3 +1638,488 @@ export default shadergenerator;
+ *   myShader = baseMaterialShader().modify(() => {
+ *     getPixelInputs(inputs => {
+ *       // Animate opacity based on x position
+ *       let t = uniformFloat(() => millis());
```
I think the uniform may need to be declared inside of `modify` but outside of `getPixelInputs`.

davepagurek commented on src/webgl/ShaderGenerator.js:
```diff
@@ -1638,3 +1638,488 @@ export default shadergenerator;
+ *     getColor((inputs, canvasContent) => {
+ *       // Warp the texture coordinates for a wavy effect
+ *       let warped = [inputs.texCoord.x, inputs.texCoord.y + 0.1 * sin(inputs.texCoord.x * 10.0)];
+ *       return getTexture(canvasContent, warped);
```
This one looks great, nice work!

Abhayaj247 replied:
Thank you! I'm glad the getColor example is working well.

davepagurek commented on src/webgl/ShaderGenerator.js:
```diff
@@ -1638,3 +1638,488 @@ export default shadergenerator;
+ *     getObjectInputs(inputs => {
+ *       // Create a sine wave along the x axis in object space
+ *       let t = uniformFloat(() => millis());
+ *       inputs.position.y += 20 * sin(inputs.position.x * 0.1 + t * 0.002);
```
In object space, I think the coordinates of the sphere are between -1 and 1, so the offset by 20 ends up looking pretty big. Maybe try with smaller values, like 0.5?

Abhayaj247 replied:
You're absolutely right! I've reduced the sine wave offset from 20 to 0.5 for the getObjectInputs example. This makes much more sense for the -1 to 1 coordinate range in object space.

davepagurek commented on src/webgl/ShaderGenerator.js:
```diff
@@ -1638,3 +1638,488 @@ export default shadergenerator;
+ *   myShader = baseColorShader().modify(() => {
+ *     getObjectInputs(inputs => {
+ *       // Create a sine wave along the x axis in object space
+ *       let t = uniformFloat(() => millis());
```
Same as before, let's move this declaration up out of `getObjectInputs` but still inside of `modify`

Abhayaj247 replied:
Done! I've moved the uniformFloat declaration outside of the getObjectInputs callback but still inside the modify function, just like with getPixelInputs.

davepagurek commented on src/webgl/ShaderGenerator.js:
```diff
@@ -1638,3 +1638,488 @@ export default shadergenerator;
+ *     getPixelInputs(inputs => {
+ *       // Animate opacity based on x position
+ *       let t = uniformFloat(() => millis());
+ *       inputs.opacity = 0.5 + 0.5 * sin(inputs.texCoord.x * 10.0 + t * 0.002);
```
Taking a look at the error in the console, I don't think `opacity` exists in the material shader's inputs:
<img width="777" height="292" alt="image" src="https://github.com/user-attachments/assets/911317d5-a0c5-4641-9ec8-3582713febcd" />

Nor the stroke shader:
<img width="809" height="365" alt="image" src="https://github.com/user-attachments/assets/b32bef34-a8be-44cc-bc93-33cd09af2eeb" />

So we'll need to update this callback to use a property that does exist. Also, we should probably update the docs to have bullet lists of properties for both the material and stroke shaders, since they have different properties.

Abhayaj247 replied:
Thanks for flagging that! You’re right - opacity isn’t part of the material or stroke shader Inputs.

I’ve updated the `getPixelInputs` example to use valid fields. It now animates the alpha channel of `inputs.color` based on `texCoord.x`, and moves the uniform outside the callback:

```

myShader = baseMaterialShader().modify(() => {
  let t = uniformFloat(() => millis());
  getPixelInputs(inputs => {
    if (inputs.color && inputs.texCoord) {
      inputs.color.a = 0.5 + 0.5 * sin(inputs.texCoord.x * 10.0 + t * 0.002);
    }
    return inputs;
  });
});

```

I also expanded the docs to list the available properties for callbacks (e.g., `texCoord`, `color`, etc.) so it’s clear which fields are valid for each shader. Let me know if you’d like me to add any additional fields or clarify further.

Abhayaj247 commented:
> Looking at the `combineColors` function, I see this error in the console: <img alt="image" width="1490" height="1036" src="https://private-user-images.githubusercontent.com/5315059/476041965-f8466317-c0ed-4a6d-a4cf-c295e617d242.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTQ2NjIwMTQsIm5iZiI6MTc1NDY2MTcxNCwicGF0aCI6Ii81MzE1MDU5LzQ3NjA0MTk2NS1mODQ2NjMxNy1jMGVkLTRhNmQtYTRjZi1jMjk1ZTYxN2QyNDIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDgwOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA4MDhUMTQwMTU0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MDE4ZjM5MzYyZmRhMTlkZmEzNGViZjFjYjY4MmE5Mzg2NWIzOGM3YWM0MjQ3ZTBjMTU3NDRjYmE2MjlhODA1NCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.noteX99OEeDOgi8foaSdStva3Bn0znp3jOGmNiHwyFc">
> 
> I think the problem for this one is that we're combining some _scalar_ values (like `baseColor.r`) with some _vector_ values (`components.ambient`):
> 
> ```js
> let r = components.baseColor.r * components.diffuse +
>               components.ambientColor.r * components.ambient +
>               components.specularColor.r * components.specular +
>               components.emissive.r + 0.2;
> ```
> 
> I think to make this work, we need to consistently use one or the other. So that could mean also doing `components.diffuse.r`, `components.specular.r`, etc. Or alternatively, doing the whole combination as a vector operation, e.g.:
> 
> ```js
> return [
>         components.baseColor * components.diffuse +
>           components.ambientColor * components.ambient +
>           components.specularColor * components.specular +
>           components.emissive +
>           [0.2, 0, 0], // Your extra tint
>         components.opacity
>       ];
> ```
> 
> (We may also want to make the tint not be in the red channel since the fill in that example is already red, making the tint harder to notice.)
> 
> I haven't looked at the other ones that don't show up just yet but that's my process for figuring out the issue: opening the console and seeing what the error is and then going from there.
> 
> Also for what it's worth, the ones that you mentioned appeared but not as 3D do show up in the 3D -> Material section for me, is that what you were referring to? Or do you just mean that it's drawing 2D content? <img alt="image" width="1470" height="1035" src="https://private-user-images.githubusercontent.com/5315059/476044833-f579188a-1c49-44af-941a-9f6b5a1f899f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTQ2NjIwMTQsIm5iZiI6MTc1NDY2MTcxNCwicGF0aCI6Ii81MzE1MDU5LzQ3NjA0NDgzMy1mNTc5MTg4YS0xYzQ5LTQ0YWYtOTQxYS05ZjZiNWExZjg5OWYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDgwOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA4MDhUMTQwMTU0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OGNmMGMyMjU5OGUxZGE4ODBlNzJhZjRjNWNmMDE0NzcyODIxYzE2Zjk5NGI5NWU2ZWJiYWRlMWU0Y2MxMDYzMSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.MWyHr_jdPnYtVYShvWmM-dA7Lpi4tKntS3oKbX6aZeE">

**Thank you for the detailed feedback @davepagurek !**

- **For `combineColors`, I see the issue with mixing scalars and vectors. I’ll update the example to use vector operations throughout, as you suggested, and will pick a more visible tint color.**
- **I’ll also review the other hooks that aren’t appearing by checking for errors in the browser console and fixing any similar issues.**
- **Regarding that appeared not as 3D the hooks `getFinalColor` and `getColor`: they do appear under the 3D → Material section in the reference, but the examples themselves are rendering as 2D shapes (like a flat circle or heart), not as 3D objects. Is this the intended behavior, or should these examples be updated to use 3D primitives for consistency with the section?**

I’ll make these updates and follow up with any further questions or fixes. Thanks again for your guidance!

Abhayaj247 commented:
Hi @davepagurek 

**Quick check on shouldDiscard: I changed the example to the string-based form to avoid the earlier runtime error.**

```

myShader = baseStrokeShader().modify({
  'bool shouldDiscard': '(bool outside) { return outside; }'
});

```

**It renders correctly (3D → Stroke) with no console errors. Is this usage/signature what you’d like to show, or would you prefer a version that produces a more visible effect (e.g., always keep: `return false`;, or always discard: `return true`;)? Happy to adjust.**

📸 shouldDiscard
Attached Screenshot/Video below:
<img width="1072" height="507" alt="shouldDiscard" src="https://github.com/user-attachments/assets/91696c8a-b212-42e3-8dcf-de3ef8316c1a" />

Abhayaj247 commented:
**Thanks for the detailed review! @davepagurek , I’ve addressed each point:**

- 1) Uniform scope: Moved the `uniformFloat(() => millis())` declarations inside modify() but outside the callbacks (both `getPixelInputs` and `getObjectInputs`).
- 2) getFinalColor docs: Updated to say “a four component vector representing red, green, blue, and alpha” for consistency.
- 3) getFinalColor example: Removed `color.a = 1` and switched to a clearer change: `color.b += 0.2.`
- 4) getColor example: Left as-is (thanks!).
- 5) getObjectInputs amplitude: Reduced from `20` to `0.5` to match the -1..1 object-space range.
- 6) getObjectInputs uniform scope: Moved the uniform declaration up into `modify()` (same pattern as #1).
- 7) getColor docs (properties): Expanded into a nested bullet list with clear descriptions (`texCoord`, `canvasSize`, `texelSize`, `color`, and what they represent).
- 8) getPixelInputs field + docs: Replaced `inputs.opacity` (not a field) with `inputs.color.a` animation. Also updated docs to list properties for both Material and Stroke shaders, so users see exactly what’s available.

**Key example updates:**

```

// getPixelInputs
myShader = baseMaterialShader().modify(() => {
  let t = uniformFloat(() => millis());
  getPixelInputs(inputs => {
    if (inputs.color && inputs.texCoord) {
      inputs.color.a = 0.5 + 0.5 * sin(inputs.texCoord.x * 10.0 + t * 0.002);
    }
    return inputs;
  });
});

// getObjectInputs
myShader = baseColorShader().modify(() => {
  let t = uniformFloat(() => millis());
  getObjectInputs(inputs => {
    inputs.position.y += 0.5 * sin(inputs.position.x * 0.1 + t * 0.002);
    return inputs;
  });
});

// getFinalColor
getFinalColor(color => {
  color.b += 0.2;
  return color;
});


```

**I've made all the requested changes. If you have a moment, could you please review and let me know if any further adjustments are needed?**

**Thank you for your time!**

davepagurek commented on src/webgl/ShaderGenerator.js:
```diff
@@ -1638,3 +1638,503 @@ export default shadergenerator;
+ *     let t = uniformFloat(() => millis());
+ *     getPixelInputs(inputs => {
+ *       // Animate alpha (transparency) based on x position
+ *       if (inputs.color && inputs.texCoord) {
```
These properties will always exist, we don't need an if statement here.

Abhayaj247 replied:
Removed the `if` check for `inputs.color` and `inputs.texCoord` as these properties are always present.

davepagurek commented on src/webgl/ShaderGenerator.js:
```diff
@@ -1638,3 +1638,503 @@ export default shadergenerator;
+ *     getWorldInputs(inputs => {
+ *       // Move the vertex up and down in a wave in world space
+ *       // In world space, moving the object (e.g., with translate()) will affect these coordinates
+ *       inputs.position.y += 0.5 * sin(t * 0.001 + inputs.position.x * 0.05);
```
I think maybe the wrong one got updated -- a scale of 20 makes sense here because in world space, the sphere is 50 units tall. The update to change the scale to 0.5 was for *object space* where the sphere is 2 units tall.
```suggestion
 *       inputs.position.y += 20 * sin(t * 0.001 + inputs.position.x * 0.05);
```

Abhayaj247 replied:
Changed the wave displacement scale from `0.5` to `20` for the world-space example, matching the object’s actual size.

davepagurek commented on src/webgl/ShaderGenerator.js:
```diff
@@ -1638,3 +1638,503 @@ export default shadergenerator;
+/**
+ * @method shouldDiscard
+ * @description
+ * Registers a callback to decide whether to discard (skip drawing) a fragment (pixel) in the fragment shader. This hook can be used inside <a href="#/p5/baseStrokeShader">baseStrokeShader()</a>.modify() and similar shader modify calls to create effects like round points or custom masking. The callback receives a boolean:
```
I think we talked about marking this one as `@private`, can we do that?

Abhayaj247 commented on src/webgl/ShaderGenerator.js:
```diff
@@ -1638,3 +1638,503 @@ export default shadergenerator;
+/**
+ * @method getWorldInputs
+ * @description
+ * Registers a callback to modify the world-space properties of each vertex in a shader. This hook can be used inside <a href="#/p5/baseColorShader">baseColorShader()</a>.modify() and similar shader modify calls to customize vertex positions, normals, texture coordinates, and colors before rendering. "World space" refers to the coordinate system of the 3D scene, before any camera or projection tr ...
```
Added a parameters bullet list for `getWorldInputs`, matching the format and content from `getObjectInputs`/`getCameraInputs`.

Abhayaj247 commented:
**Thanks for the detailed review! @davepagurek , I’ve addressed each point:**

1. **getColor docs:** Removed the `color` bullet from the parameter list since it’s not an actual input property.
2. **getPixelInputs example:** Removed the unnecessary `if` checks for `inputs.color` and `inputs.texCoord`, as these are always present.
3. **Shape example:** Updated `sphere(50)` to `circle(0, 0, 100)` to avoid 3D transparency artifacts in the example.
4. **getWorldInputs wave example:** Changed the displacement scale from `0.5` to `20` to better match the object’s actual size in world space.
5. **shouldDiscard method:** Added the `@private` JSDoc tag as suggested.
6. **getWorldInputs docs:** Added a parameters bullet list (`position`, `normal`, `texCoord`, `color`) to match the style of `getObjectInputs`/`getCameraInputs`.

**All requested changes have been made. Please let me know if you spot anything else that needs adjusting.**

**Thank you for your guidance!**

Abhayaj247 commented on src/webgl/ShaderGenerator.js:
```diff
@@ -1638,3 +1638,507 @@ export default shadergenerator;
+/**
+ * @method getWorldInputs
+ * @description
+ * Registers a callback to modify the world-space properties of each vertex in a shader. This hook can be used inside <a href="#/p5/baseColorShader">baseColorShader()</a>.modify() and similar shader modify calls to customize vertex positions, normals, texture coordinates, and colors before rendering. "World space" refers to the coordinate system of the 3D scene, before any camera or projection tr ...
```
Thanks for pointing this out! I’ve added a link to `.modify()` so it’s consistent with other references.

Abhayaj247 commented on src/webgl/ShaderGenerator.js:
```diff
@@ -1638,3 +1638,507 @@ export default shadergenerator;
+ *   shader(myShader);
+ *   lights();
+ *   noStroke();
+ *   fill('red');
```
Makes sense - I’ve replaced `fill('red')` with `fill('white')` so the shader’s colors show clearly.

Abhayaj247 commented:
**Hi @perminder-17 ,**

**Thanks a lot for the detailed review 🙌 I’ve gone through your comments and made the suggested changes :**

**1. Added links for `.modify()` references where mentioned.**
**2. Updated `@param {function}` to `@param {Function}` consistently.**
**3. Removed `fill('red')` and replaced with `fill('white')` so the shader colors show correctly.**
**4. Rephrased “a vector with three components” → “a three-component vector” and added periods at the end of each line for consistency.**
**5. Applied the same rephrasing and punctuation improvements in stroke-related docs.**
**6. Increased the blue tint value from `0.2` → `0.4` for better visibility.**
**7. Reduced sine wave amplitude from `20` → `3` to keep the sphere in view.**
**8. Changed `sphere(100)` → `sphere(50)` so it fits comfortably in the 200×200 canvas.**

**All requested changes have been made. Please let me know if you spot anything else that needs adjusting.**

**Thank you for your guidance!**

davepagurek commented on src/webgl/ShaderGenerator.js:
```diff
@@ -1638,3 +1638,509 @@ export default shadergenerator;
+ *   background(200);
+ *   shader(myShader);
+ *   noStroke();
+ *   fill('blue');
```
Currently the `inputs.color.b = 1` has no effect if you comment it out because the input color's blue channel is already maxed out from the `fill('blue')` here. Can we use `fill('red')` so that we get a visible color change?

davepagurek commented on src/webgl/ShaderGenerator.js:
```diff
@@ -1638,3 +1638,509 @@ export default shadergenerator;
+ *   shader(myShader);
+ *   noStroke();
+ *   fill('orange');
+ *   sphere(100);
```
@perminder-17 mentioned it on another example, but on here too `sphere(100)` fills the canvas entirely, I think `sphere(50)` leaves some room for it to go up and down:

https://github.com/user-attachments/assets/ee838bc6-eceb-444f-9a0d-2208cc7bb3af

Abhayaj247 commented on src/webgl/ShaderGenerator.js:
```diff
@@ -1638,3 +1638,509 @@ export default shadergenerator;
+ *     let t = uniformFloat(() => millis());
+ *     getObjectInputs(inputs => {
+ *       // Create a sine wave along the x axis in object space
+ *       inputs.position.y += 3 * sin(t * 0.001 + inputs.position.x * 0.05);
```
Makes sense - I’ve simplified the sine wave expression by removing the multipliers as suggested. It now uses:

`inputs.position.y += sin(t * 0.001 + inputs.position.x); `

perminder-17 commented:
@all-contributors please add @Abhayaj247 for doc

Abhayaj247 commented:
Hi @perminder-17 , thank you again for merging my PR and for adding me as a contributor 🙏.

I was just curious about something - I noticed the PR doesn’t appear under my “Merged Pull Requests” on my GitHub profile. Could this be related to the merge method (like squash merges not preserving commit authorship)? Not an issue at all, but I wanted to understand how it works.

Really appreciate your guidance and support throughout this process!

Abhayaj247 commented:
> I think we merged it on the `dev-2. 0` branch which is not the default branch, maybe that's the reason your commits are not showing up.

Hi @perminder-17 , thanks for clarifying! 🙏
That explains why my commits aren’t showing under “Merged Pull Requests.” I really appreciate your help and guidance throughout this process.
